File: day08/day8.py
BASE_MAP = {2: 1,  3: 7,  4: 4,  7: 8}
# 1 => 2
# 7 => 3
# 4 => 4
# 2 => 5, 3 => 5, 5 => 5
# 0 => 6, 6 => 6, 9 => 6
# 8 => 7

def part1(data):
    return sum(sum(1 for ent in out if len(ent) in BASE_MAP.keys()) for _,out in data)

def strsrt(s):
    return ''.join(sorted(s))

def isct(s1, s2):
    return ''.join(c for c in s1 if c in s2)

def getBaseFour(data):
    return {ent:BASE_MAP[len(ent)] for ent in data if len(ent) in BASE_MAP.keys()}

def decode(inp):
    nums = getBaseFour(strsrt(e) for e in inp)
    lets = {v:k for k,v in nums.items()}
    for ent in inp:
        ent = strsrt(ent)
        if len(ent) == 5:
            if len(isct(lets[1], ent)) == 2: nums[ent] = 3
            elif len(isct(lets[4], ent)) == 3: nums[ent] = 5
            else: nums[ent] = 2
        elif len(ent) == 6:
            if len(isct(lets[1], ent)) != 2: nums[ent] = 6
            elif len(isct(lets[4], ent)) == 4: nums[ent] = 9
            else: nums[ent] = 0
    return nums

def numerize(out, decoded):
    return sum(10**(3-i) * decoded[strsrt(n)] for i,n in enumerate(out))

def part2(data):
    return sum(numerize(out, decode(inp)) for inp,out in data)

def main(fname):
    with open(fname) as f:
        data = [x.split(' | ') for x in f.read().strip().split('\n')]
    data = [(a.split(' '), b.split(' ')) for (a,b) in data]

    print(f'Part 1: {part1(data)}')
    print(f'Part 2: {part2(data)}')

if __name__ == "__main__":
    import sys
    filename = 'input.txt' if len(sys.argv) <= 1 else sys.argv[1] 
    main(filename)

File: day08/day8.exs
defmodule Solve do
    # maps num_segments -> number
    @base_segments %{2 => 1, 3 => 7, 4 => 4, 7 => 8}

    def read_input(fname) do
        data = File.read!(fname)
            |> String.trim
            |> String.split("\n")
            |> Enum.map(&(String.split(&1, " | ")))
        for [left, right] <- data, do: [String.split(left, " "), String.split(right, " ")]
    end

    def part1(data) do
        base4s = for entry <- data,
            outputs = entry |> Enum.at(1),
            signal <- outputs,
            String.length(signal) in Map.keys(@base_segments),
            do: signal
        base4s |> length 
    end

    def part2(data) do
        entries = for entry <- data,
            [inps, outs] = entry,
            base_four = get_base_four(inps),
            decoded = decode_inputs(base_four, inps),
            do: decode_output(decoded, outs)
        entries |> Enum.sum
    end

    defp get_base_four(from_input) do
        for signal <- from_input |> Enum.map(&(sort(&1))) |> Enum.uniq,
            len = String.length(signal),
            len in Map.keys(@base_segments),
            into: %{},
            do: {signal, @base_segments[len]}
    end

    defp decode_inputs(codes, input) do
        rev_codes = for {segs, num} <- codes, into: %{}, do: {num, segs}
        decode_input(input, codes, rev_codes)
    end

    defp decode_input([head | tail], codes, rev_codes) do
        head = sort(head)
        codes = case String.length(head) do
            5 -> cond do
                    isect_len(rev_codes[1], head) == 2 ->
                        Map.put(codes, head, 3)
                    isect_len(rev_codes[4], head) == 3 ->
                        Map.put(codes, head, 5)
                    true -> Map.put(codes, head, 2)
                end
            6 -> cond do
                    isect_len(rev_codes[1], head) != 2 ->
                        Map.put(codes, head, 6)
                    isect_len(rev_codes[4], head) == 4 ->
                        Map.put(codes, head, 9)
                    true -> Map.put(codes, head, 0)
                end
            _ -> codes
            end
        decode_input(tail, codes, rev_codes)
    end
    defp decode_input([], codes, _), do: codes

    defp decode_output(codes, output) do
        digits = for {signal, i} <- output |> Enum.map(&(sort(&1))) |> Enum.with_index,
            val = codes[signal],
            do: val * 10 ** (3-i)
        digits |> Enum.sum
    end

    defp sort(str) do
        str |> String.to_charlist
            |> Enum.sort
            |> List.to_string
    end

    defp isect_len(str1, str2) do
        ar1 = str1 |> String.to_charlist
        ar2 = str2 |> String.to_charlist
        ar1 -- (ar1 -- ar2) |> length
    end
end

data = Solve.read_input("input.txt")

IO.puts "Part 1: #{Solve.part1(data)}"
IO.puts "Part 2: #{Solve.part2(data)}"